proposals/000-decorator-api.md

```diff
@@ -0,0 +1,322 @@
+* **Decoration style**: Determines the look and feel of a decoration once rendered by a Navigator. For example `highlight`, `underline` or `margin icon`. It is media type agnostic.
+* **Decoration template**: Translates a decoration and its style into concrete rendering instructions for the Navigator. Template implementations depend on the resource media type and the particular Navigator used.
+
+### Creating a New Decoration
+
+The `Decoration` object associates a location (`Locator`) in a publication with a *decoration style* to render. By relying on the style to indicate its look and feel, `Decoration` avoids any specific rendering instructions. This allows to reuse the same decorating code for any kind of publication or different Navigators.
+
+Each `Decoration` object has an identifier which **must** be unique in the decoration's group. This is necessary to compute efficiently changes when applying updates. A database ID is a perfect candidate.
+
+```swift
+Decoration(
+    id: "12",
+    locator: Locator(...),
+    style: Decoration.Style.MarginIcon(image: "bookmark")
+)
+```
+
+Note that a single entity might be rendered using several `Decoration` objects. For example, a highlight with a text annotation needs two `Decoration` objects: one for the actual highlight and one to draw the "text annotation" icon in the page margin.
+
+### Applying Decorations to a Navigator
+
+A reading app never explicitly asks the Navigator to render a decoration. Instead, the app declares the current state of the decorations for a given *decoration group* and the Navigator will decide when to render each decoration. Navigators use a diff algorithm to determine the actual changes since the last applied state to run efficient rendering instructions.
+
+```swift
+// First, fetch highlight models from your database.
+let highlights = ...
+
+// Then, create one new decoration per highlight.
+let decorations = highlights.map { highlight in
+    Decoration(
+        id: highlight.id,
+        locator: highlight.locator,
+        style: Decoration.Style.Highlight(tint: highlight.color)
+    )
+}
+
+// Finally, apply the decorations in a group named "user-highlights".
+navigator.apply(decorations: decorations, in: "user-highlights")
+```
+
+### Handling User Clicks on Decorations
+
+To allow user clicks/taps on a decoration, you must first check that the `activate` interaction is available for the decoration style:
+
+```swift
+navigator.supportsDecorationInteraction(.activate, forStyle: Decoration.Style.Highlight)
+```
+
+Then, enable the `activate` interaction when creating each `Decoration` object.
+
+```swift
+Decoration(
+    id: "12",
+    locator: Locator(...),
+    style: Decoration.Style.Highlight(),
+    interactions: [.activate]
+)
+```
+
+Finally, listen to the `DecorationActivated` event which will be emitted when the user clicks the decoration. The API depends on the platform.
+
+```swift
+navigator.observe(DecorationActivated) { decoration in
```

Comment:
We might offer the ability to observe interactions on only a specific decoration group. Maybe at the same time as we provide the decoration list. For instance, with something like `navigator.registerDecorationGroup(decorations: Observable<List>, callback: (Decoration) -> Unit)`

Reply:
Yeah, something to be explored in the Kotlin implementation I think. Traditionally with Swift it's implemented with a delegate method so there's no registration and it all goes into the same place. But it would make sense to have per-group observers/listeners.

```swift
func navigator(_ navigator: DecorableNavigator, didActivateDecoration decoration: Decoration, in group: String)
```